File: google/main/src/core/query/OrTerms.java
package main.src.core.query;

import java.util.regex.Pattern;
import java.util.*;

import main.src.core.engine.InvertedIndex;
import main.src.core.structures.Token;
import main.src.core.structures.Document;


public class OrTerms extends Terms{

    private static Pattern pattern = Pattern.compile("\\+(\\w+)");

    public OrTerms(String expression){
        super();
        this.collect(expression, pattern, 1);
    }

    public HashSet<Document> getResults(InvertedIndex index){
        HashSet<Document> results = new HashSet<>();
        for(Token token : this.keys)
            results.addAll(index.getDocumentsOfToken(token));
        return results;
    }
}

File: GoogleSharp/GoogleSharp/src/core/query/OrTerms.cs
using System;
using System.Linq;
using GoogleSharp.Src.Core.Structures;
using GoogleSharp.Src.Core.Engine;
using System.Text.RegularExpressions;
using System.Collections.Generic;

namespace GoogleSharp.Src.Core.Query
{
    public class OrTerms : Terms
    {
        private static readonly Regex Pattern = new Regex(@"\+(\w+)", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        public OrTerms(String expression) : base()
        {
            this.Collect(expression, Pattern, 1);
        }

        public override HashSet<Document> GetResults(IInvertedIndex index)
        {
            return this.Tokens.SelectMany(token => index.GetDocumentsOfToken(token)) as HashSet<Document>;
        }
    }
}